### DISCUSSION TITLE
failover  support of aurora

### DISCUSSION THREAD

**Discussion by pokerstarxy** [sqlalchemy/sqlalchemy#7849] [Usage Questions]
### Describe the use case

Hello,guys. we are using connecting pools  build with sqlalchemy.  but we have some trouble in connecting to the   aurora which is a db cluster service of  aws.
When the cluster failsover, the connection pool is not rebuilt, resulting in us getting a read-only connection. The root cause is that the aurora switchover is a change in dns resolution and the instance we are connecting to goes from a read-write state to a read-only state. The read-only instance before the switchover becomes a read-write instance after the switchover, but the database connection does not change at all. 

Is there any good way to achieve this please? 

 I hope you can give me some advice from your side

### Databases / Backends / Drivers targeted

https://aws.amazon.com/rds/aurora/?nc1=h_ls&aurora-whats-new.sort-by=item.additionalFields.postDateTime&aurora-whats-new.sort-order=desc

### Example Use

*

### Additional context

Online I found a java based implementation

https://mariadb.com/kb/en/about-mariadb-connector-j/#failover-and-load-balancing-modes

<img width="799" alt="企业微信截图_f4f64d8a-eb2c-479c-a246-17b2f442ae45" src="https://user-images.githubusercontent.com/16798673/159524414-43834941-f305-4aa4-b13c-f2a370542fc1.png">

**Comment by zzzeek** [comment]
hi there -

unfortunately this question is out of scope for the SQLAlchemy library and has more to do with how your aurora database failover is configured and what capabilities are available from the database driver you're using.  For example if you are using mariadb-connector, there may be some guidance at https://jira.mariadb.org/secure/Dashboard.jspa , although this looks more like a general Amazon AWS type of issue, if you can get support from them.

**Reply by zzzeek** [reply]
from the SQLAlchemy side, if you need to invalidate all connections in the pool, you can call `engine.dispose()` - see https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=engine%20dispose#engine-disposal .  That's about the extent of what role SQLAlchemy would have here.

**Reply by pokerstarxy** [reply]
Do you mean to check the connection status at regular intervals? Rebuild the connection pool if it doesn't match

**Reply by zzzeek** [reply]
something like that.  if this failover has connections moving from read/write to read, but yet there is somehow another "read/write" place to switch towards, why doesn't the failover move to the next read/write instance  ?   is it that there is a delay?       if these connections can't tolerate being read only I would look to see if there's a way that your live connections fail over to just not being connected, then you can use pre-ping in the connection pool so that they transparently reconnect.   if there's some period of time that no R/W database is available, and your application can't tolerate returning errors during this period, then your application would need some kind of retry logic as well, but that's probably a different issue.

If there is an inexpensive way to test a live connection for read/write or read-only status, you could also use a checkout handler to perform this test at checkout time, using a form similar to that illustrated at https://docs.sqlalchemy.org/en/14/core/pooling.html#custom-legacy-pessimistic-ping .    you can .invalidate() a connection if you deem it not usable, like:

```py

import time
from sqlalchemy import event
from sqlalchemy import create_engine

some_engine = create_engine("...")

@event.listens_for(some_engine, "engine_connect")
def ping_connection(connection, branch):
    if branch:
        # "branch" refers to a sub-connection of a connection,
        # we don't want to bother pinging on these.
        return

    for attempt in range(3):
        is_read_only= connection.exec_driver_sql("SELECT read_only FROM something").scalar()
        if is_read_only:
            connection.invalidate()
            time.sleep(1)  # wait 1 second before trying again
        else:
            break
    else:
        raise Exception("connection stayed read only for 3 attempts")
```

**Reply by methane** [reply]
Since there are cases where an existing connection becomes read_only, we should listen to the Pool.checkout event.

```python
@event.listens_for(engine, "checkout")
def handle_pool_checkout(dbapi_connection, connection_record, connection_proxy):
    try:
        cursor = dbapi_connection.cursor()
        cursor.execute("SELECT @@innodb_read_only")
        read_only = cursor.fetchone()[0]
        if read_only:
            raise DisconnectionError("detected innodb_read_only=1")
    except Exception as e:
        # We can use this query as a replacement for pool_pre_ping.
        raise DisconnectionError(str(e))
```
